ocaml/quicktest/qt_filter.ml

```diff
@@ -0,0 +1,200 @@
+
+module A = Quicktest_args
+
+type 'a test_case = string * Alcotest.speed_level * 'a
+
+type ('a, 'b) filter = 'a test_case list -> 'b test_case list
+
+let for_each f test_cases = List.map f test_cases |> List.concat
+(** Creates a filter from a function that processes only one test case,
+    by applying this function to every test case in the list, and concatenating
+    the results. *)
+
+let session_id = ref API.Ref.null
+
+let vdi_count = Hashtbl.create 4
+
+(** Scan an SR and return the number of managed VDIs contained within *)
+let count_vdis rpc session_id sr =
+  Client.Client.SR.scan rpc session_id sr;
+  let managed_vdis =
+    Client.Client.SR.get_VDIs rpc session_id sr
+    (* NB vhd backends may delete records beneath us *)
+    |> Valid_ref_list.filter
+      (fun vdi -> Client.Client.VDI.get_managed rpc session_id vdi)
+  in
+  List.length managed_vdis
+
+(** Called before the quicktests start to save the original state.
+    This data will be used by [finish] to check for any resouce leaks. *)
+let init () =
+  session_id := Qt.init_session !A.rpc !A.username !A.password;
+  Client.Client.SR.get_all_records ~rpc:!A.rpc ~session_id:!session_id
+  |> List.iter (fun (ref, sr) ->
+    if List.mem `scan sr.API.sR_allowed_operations then begin
+      let before = count_vdis !A.rpc !session_id ref in
+      Hashtbl.add vdi_count sr.API.sR_uuid before;
+      Printf.printf "Hashtbl.add vdi_count %s %d\n%!" sr.API.sR_uuid before
+    end)
+
+(** Called at the end of the quicktests to check that no resources leaked
+    during the test run *)
+let finish () =
+  Client.Client.SR.get_all_records ~rpc:!A.rpc ~session_id:!session_id
+  |> List.iter (fun (ref, sr) ->
+      Printf.printf "Hashtbl.find_opt vdi_count %s\n%!" sr.API.sR_uuid;
+      match Hashtbl.find_opt vdi_count sr.API.sR_uuid with
+      | Some before ->
+          Printf.printf "Hashtbl.find_opt vdi_count %s = %d\n%!" sr.API.sR_uuid before;
+          if List.mem `scan sr.API.sR_allowed_operations then begin
+            let after = count_vdis !A.rpc !session_id ref in
+            if after <> before then
+              failwith (Printf.sprintf "VDIs leaked on SR %s: before=%d, after=%d" sr.API.sR_uuid before after)
```

Comment:
`Alcotest.failf`?

Reply:
This is currently run outside alcotest - I might have a followup PR to convert this into a test case.

Reply:
Ah okay, I'll stop being pedantic then :P